<a href="https://colab.research.google.com/github/umass-userid/data-science-for-business/blob/main/Guided_Case_2_Telco_Customer_Churn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Case 2: Telco Customer Churn
"Predict behavior to retain customers. You can analyze all relevant customer data and develop focused customer retention programs." [IBM Sample Data Sets](
https://community.ibm.com/community/user/businessanalytics/blogs/steven-macko/2019/07/11/telco-customer-churn-1113)

### Import the data
The URL is a Google Drive link set to allow anyone with the link to view. This is a useful block of code to keep nearby. The only piece that changes is the value of the "orig_url".   

In [ ]:
import pandas as pd
import requests
from io import StringIO

orig_url="https://drive.google.com/file/d/18e7X8-KyzlX6K1P8sujnVDVtERgsWD9P/view?usp=sharing"

file_id = orig_url.split('/')[-2]
dwn_url='https://drive.google.com/uc?export=download&id=' + file_id
url = requests.get(dwn_url).text

csv_raw = StringIO(url)
telco = pd.read_csv(csv_raw)
telco.head()

Account_Length  Vmail_Message  Day_Mins  ...  State  Area_Code     Phone
0             128             25     265.1  ...     KS        415  382-4657
1             107             26     161.6  ...     OH        415  371-7191
2             137              0     243.4  ...     NJ        415  358-1921
3              84              0     299.4  ...     OH        408  375-9999
4              75              0     166.7  ...     OK        415  330-6626

[5 rows x 21 columns]

### Import some useful libraries

In [ ]:
import numpy as np # array operations
import pandas as pd # data processing
import matplotlib.pyplot as plt # visuals
import seaborn as sns # visuals
import warnings
warnings.filterwarnings("ignore") # suppress warnings for readability

### 1. Data Exploration

#### 1.a. Determine the shape of the data.

In [ ]:
# find the shape of the telco data set.
telco.---

(3333, 21)

#### 1.b. What do the numbers ```(3333, 21)``` tell you?
3333 rows     
21 columns (features, attributes)


#### 1.c. Deterimine whether or not there are any missing values.

In [ ]:
telco.----().sum()

#### 1.d. Determine the count of each of the Churn  (target variable) categories ('yes', 'no'). Use the ```value_counts()``` method.

In [ ]:
# separate churn
telco['Churn'].----

#### 1.e. Create a histogram of the 'Account_Length'. Use the Seaborn `distplot()` method.

In [ ]:
import matplotlib as plt
import seaborn as sns
sns.----(telco['Account_Length'])

#### 1.f. Create a boxplot of the Account_Length by Churn status.

In [ ]:
sns.----(x='Churn', y='Account_Length', data=telco)

### 1.g. Describe any patterns in Churn based on tenure.
(double click to edit)














### 2. Data Preprocessing

#### 2.a. Determine the data types for all of the variables in the telco data set. Use the `dtypes` attribute.

In [ ]:
telco.----

#### 2.b. Based on the results above, which 5 variables require modification before a model can be built from this data set?
(double click to edit)

What do the ```Churn``` values look like?

In [ ]:
telco['Churn'].head()

Replace all of the ```'no'``` values with 0 and all of the ```'yes'``` values with 1.

In [ ]:
telco['Churn']=telco['Churn'].replace({'no':0, 'yes':1})
telco['Churn'].head()

What do the ```'Intl_Plan'``` values look like?

In [ ]:
telco['Intl_Plan'].head()

Replace all of the ```'no'``` values with 0 and all of the ```'yes'``` values with 1.

In [ ]:
telco['Intl_Plan']=telco['Intl_Plan'].replace({'no':0, 'yes':1})
telco['Intl_Plan'].head()

#### 2.c. What do the ```'Vmail_Plan'``` values look like? Examine the first few rows with the `head()` method.

In [ ]:
telco['Vmail_Plan'].----()

#### 2.d. Replace all of the ```'no'``` values with 0 and all of the ```'yes'``` values with 1 for the ```'Vmail_Plan'``` data.

In [ ]:
telco['Vmail_Plan']=telco['Vmail_Plan'].----
telco['Vmail_Plan'].head()

Re-examine the data types of the telco data.

In [ ]:
telco.dtypes

What do the ```'State'``` values look like?

In [ ]:
telco['State'].head()

Use the Pandas ```get_dummies``` method to create a column of indicators for each state.

In [ ]:
telco=pd.get_dummies(data=telco, columns=['State'], drop_first=True)
telco.head()

#### 2.e. Remove the `Phone` variable from the telco data. Use the Pandas `drop()` method.

In [ ]:
# drop Phone
telco.----(['Phone'], axis=1, inplace=True)


Examine the first few rows of the data set.

In [ ]:
telco.head()

### 3. Attribute Selection (abbreviated)

Find the correlations for the quantitative variables.

In [ ]:
corr=telco[['Account_Length','Vmail_Message','Day_Mins','Eve_Mins','Night_Mins','Intl_Mins','CustServ_Calls','Day_Calls','Day_Charge','Eve_Calls','Eve_Charge','Night_Calls','Night_Charge','Intl_Calls','Intl_Charge']].corr()
corr

#### 3.a. Create a heatmap of the correlations using the Seaborn `heatmap()` method.

In [ ]:
import seaborn as sns
sns.----(corr)

#### 3.b. Based on the correlation analysis, which variables appear to be the most closely associated?
(double click to edit)

#### 3.c. What could be done to eliminate redundancy in the data?
(double click to edit)

### 4. Building the Model

Separate the collection of feature variables from the target variable.

#### 4.a. Use the Pandas `drop()` method to remove the `Churn` data from the telco data set and store the results as `features`.

In [ ]:
features=telco.----(['Churn'], axis=1)

#### 4.b. Store the telco Churn data as `target`.

In [ ]:
target=----['----']

#### 4.c. Create the four data sets `X_train, X_test, y_train, y_test` using the `train_test_split` method from `sklearn`.

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = ----(----, ----, test_size=0.2, random_state=42)


#### 4.d. Build a Logistic regression model by fitting this model to the X and y training data.

In [ ]:
# Import LogisticRegression
from sklearn.linear_model import LogisticRegression

# Instantiate the classifier
clf = LogisticRegression()

# Fit the classifier
clf.----(----, ----)

#### 4.e. Use the classifier to predict the target values based on the X testing data.

In [ ]:
# Make predictions using the collection of featuress reserved in X_test
clf.----(----)

### 5. Model Evaluation

#### 5.a. Use the `score()` method with the testing data (`X_test` and `y_test`) to determine the model's accuracy.

In [ ]:
clf_score=clf.----(----, ----)
print(clf_score)

#### 5.b. Briefly interpret the score of this classifier.
(double click to edit)

### 6. Model Tuning
This section is optional. You are welcome to experiment with the various methods for hyperparameter tuning covered in the last chapter of the DataCamp course. [link to slides](https://drive.google.com/file/d/1kgbvlrgNw30RWl6nWnyjj3aB6vOZwlL2/view?usp=sharing)